### ISSUE TITLE
Unable to run multiple geckodriver procesess in parallel

### ISSUE THREAD

**Issue by TomNelsonHBC** [mozilla/geckodriver#388]
## Firefox Version
50.0.2 and 45.0.1

## Platform
Windows Enterprise 7 64, Windows 8.1 Enterprise
C# Visual Studio 2012
Selenium 3.0.1
geckodriver 0.11.1 Win32

## Steps to reproduce
Windows 7, C#, Selenium 3.0.1, .Net Assembiles, geckodriver 0.11.1, Firefox 50.0.2
Note that with selenium 2.51 and Firefox 45.0.1 I can run up to 5+ local sessions at the same time
- Execution thread
- Create Firefox options and Set Prefrences
- Set option use legacy implementation to false 
- Create Firefox service with path \ to \ geckodriver.exe
- Set service Firefox service path to installed fire fox path
- create web driver, oWebDriver is declared as IWebDriver 
- oWebDriver = new FirefoxDriver(ffService, ffOption, tMarionette);
- With this thread executing test 
- I start another execution thread that repeats above steps
- once the second thread starts up the browser 1st browser crashes
- The first browser just says  Bad news first: This tab has crashed
- The geckodriver output 1st thread only has connection information
- The listening port and the Firefox host port  are different in the two output windows
- The profile listed it both output windows are different rust_mozprofile.SomeKey
- Each thread has a different profile, 2 geckodrivers are running, 2 Firefox browser are running
- It seems like when the second browser is started it is not started as a -new-instance

-Note when I run against Firefox 45.0.1 with use legacy implementation true I am able to run 6 threads

Note the geckodriver does not crash just the Firefox browser, but when the new instance starts the original browser flashes and then crashes, if click 1st browsers restore this tab button fast enough I get both threads to execute.  

I can not copy the trace information from the geckodriver.exe Win32 CMD window. if there is a way to output this info to a log file please let me know and I will do so.

[htmlInfoForGeckodriver.txt](https://github.com/mozilla/geckodriver/files/642661/htmlInfoForGeckodriver.txt)

**Comment by andreastt** [comment]
It may seem like we want to append `--new-instance` to https://github.com/jgraham/rust_mozrunner/blob/master/src/runner.rs#L115 and start Marionette on a random port to support multiple instances of Firefox on the same system.

**Comment by jgraham** [comment]
This works fine for me with two geckodriver instances running on different ports:

    In [2]: requests.post("http://localhost:4444/session", json={}).json()Out[2]: 
    {u'sessionId': u'22350fc0-78ec-4cfb-b02c-3d3b029c582c',
     u'value': {u'XULappId': u'{ec8030f7-c20a-464f-9b0e-13a3a9e97384}',
      u'acceptSslCerts': False,
      u'appBuildId': u'20161209095719',
      u'browserName': u'firefox',
      u'browserVersion': u'50.1.0',
      u'command_id': 1,
      u'platform': u'LINUX',
      u'platformName': u'linux',
      u'platformVersion': u'4.4.0-57-generic',
      u'processId': 20985,
      u'proxy': {},
      u'raisesAccessibilityExceptions': False,
      u'rotatable': False,
      u'specificationLevel': 0,
      u'takesElementScreenshot': True,
      u'takesScreenshot': True,
      u'version': u'50.1.0'}}

    In [3]: requests.post("http://localhost:4445/session", json={}).json()
    Out[3]: 
    {u'sessionId': u'774085c7-3482-450f-97fb-7a257f7078eb',
     u'value': {u'XULappId': u'{ec8030f7-c20a-464f-9b0e-13a3a9e97384}',
      u'acceptSslCerts': False,
      u'appBuildId': u'20161209095719',
      u'browserName': u'firefox',
      u'browserVersion': u'50.1.0',
      u'command_id': 1,
      u'platform': u'LINUX',
      u'platformName': u'linux',
      u'platformVersion': u'4.4.0-57-generic',
      u'processId': 21079,
      u'proxy': {},
      u'raisesAccessibilityExceptions': False,
      u'rotatable': False,
      u'specificationLevel': 0,
      u'takesElementScreenshot': True,
      u'takesScreenshot': True,
      u'version': u'50.1.0'}}

    In [4]: requests.post("http://localhost:4444/session/22350fc0-78ec-4cfb-b02c-3d3b029c582c/execute/sync", '{"script": "document.body.innerHTML = \'<p>1</p>\'", "args": null}').json()
    Out[4]: {u'value': None}

    In [5]: requests.post("http://localhost:4444/session/774085c7-3482-450f-97fb-7a257f7078eb/execute/sync", '{"script": "document.body.innerHTML = \'<p>2</p>\'", "args": null}').json()
    Out[5]: 
    {u'error': u'invalid session id',
     u'message': u'Got unexpected session id 774085c7-3482-450f-97fb-7a257f7078eb expected 22350fc0-78ec-4cfb-b02c-3d3b029c582c'}

    In [6]: requests.post("http://localhost:4445/session/774085c7-3482-450f-97fb-7a257f7078eb/execute/sync", '{"script": "document.body.innerHTML = \'<p>2</p>\'", "args": null}').json()
    Out[6]: {u'value': None}`

And the two Firefox instances show the correct text.

**Comment by TomNelsonHBC** [comment]
Closing this issue
Thanks for looking into, I created separate test code to test starting multiple instances and it did work for me. Found that in our start fire fox method prior to creating the driver we were checking for FF process running  that may have been left open by a previous test. This was done by checking the processes main window handle vs handle captured after starting Firefox. In FF v50 If 2 Firefox browsers are running there are actually 4 fire fox process running and two of them have a window handle of 0 (note this is not the case in FF version 45). The default for our var that stores the window handle is 0 so we were trying to kill fire fox processes that have a handle of 0.
